Add support for displaying emotes using the terminal graphics protocol

Nogesma opened the pull request:
This PR adds support for displaying emote as discussed in #265.

I don't really know windows' equivalent of `~/.cache` so I left a `todo!()` in the `cache_path` function.

I also managed to hit a few panics during my testing but they also seem to appear on your main branch so I left them as is.

Xithrius commented:
Panics are probably due to window size, they'll be fixed with the other currently open PR.

Xithrius commented:
Perhaps the cache on windows can be stored in `LOCALAPPDATA`'s path? Or was it `APPDATA`? Haven't used windows to program in a long time lol

Xithrius commented:
Looks like windows build is aliven't, hecin windows

Nogesma commented:
Yeah, seems like it's because of the dependency on sixel-rs of pic, which I don't use anyway. I might put it behind a feature flag to make it build.
Another option would be to integrate the code I use from pic directly into this repo, as it's only a few functions, and I modified almost all of them from the base repo to suit my needs.

Nogesma commented:
> Perhaps the cache on windows can be stored in `LOCALAPPDATA`'s path? Or was it `APPDATA`? Haven't used windows to program in a long time lol

I can make it a subdirectory of the config dir, like `APPDATA/twt/cache`. I never really used windows so I don't really know.

Xithrius commented:
> Yeah, seems like it's because of the dependency on sixel-rs of pic, which I don't use anyway. I might put it behind a feature flag to make it build. Another option would be to integrate the code I use from pic directly into this repo, as it's only a few functions, and I modified almost all of them from the base repo to suit my needs.

Putting it behind a feature flag would probably be a good idea, along with integration of your code from pic.

Xithrius commented:
> > Perhaps the cache on windows can be stored in `LOCALAPPDATA`'s path? Or was it `APPDATA`? Haven't used windows to program in a long time lol
> 
> I can make it a subdirectory of the config dir, like `APPDATA/twt/cache`. I never really used windows so I don't really know.

That'll probably do fine for now.

Xithrius commented:
PR #313 was merged, so panics when window is too small should be solved now. If there are any other panics, please let me know.

Nogesma commented:
> PR #313 was merged, so panics when window is too small should be solved now. If there are any other panics, please let me know.

Okay I rebased my branch on top of your main. Didn't have any panics yet. 

I integrated most of the modified code from pic into `src/utils/kitty.rs`. 
I also added support for gifs. I still need to look into making animated webp/avif work.

Xithrius commented:
Very nice.

Xithrius commented:
Running now

Nogesma commented:
I've been using this for a few days and I noticed a few issues. 

The first one is that spaces are trimmed when wrapping messages (mgeisler/textwrap#503). For now I keep a copy of the message and keep trimming it with the message part that's in the current span, to then get it's length. It's probably not optimal but I didn't really find a way to reliably get the number of characters trimmed, so that I can work with numbers instead of cloning strings.

I also replace the emote name when parsing with `'a' * emote_width`, then replace that again after wrapping with spaces. This would probably not be needed too if spaces were not trimmed.  

Another issue is that the image rust crate panics sometimes when decoding webp images/animations (image-rs/image#1856).  Webp support is needed to enable 7tv emotes. But I can remove them until this is fixed. 

I now download the emotes in another thread, as it was making the app start quite slow. 

I rebased onto main, and squashed my commits.

Nogesma commented:
Fixed clippy errors.

Xithrius commented:
I shall begin my own testing of this within the next few days.

Nogesma commented:
I pushed an update to now scale the emote automatically to the terminal row height. 
I also had to move some code from `draw_ui()` into another function as clippy was complaining about the function being too complex.

Xithrius commented:
Although I am yet to begin testing of this implementation of emotes, the code itself seems to be in good order.

Xithrius commented:
![image](https://user-images.githubusercontent.com/15021300/230245888-13c2cc59-6b1b-45a5-8d13-a7a09f02c541.png)

It works very well.

Xithrius reviewed (approved):
I believe that implementing tmux + kitty is out of scope for this PR and probably isn't possible, so I think the current state of this implementation will do.

I'll have another person review before merging this PR, and if you don't have anything else on your mind for said PR.

Nogesma commented:
I've checked and even if I remove the env check on tmux + kitty, tmux doesn't respond to the check at https://github.com/Nogesma/twitch-tui/blob/09a282cdaa876a17dbd791b0f114aeebafe1c0ec/src/emotes/kitty.rs#L69 to see if the terminal supports the graphics protocol, and doesn't respond either to the query for the terminal size in pixels at https://github.com/Nogesma/twitch-tui/blob/09a282cdaa876a17dbd791b0f114aeebafe1c0ec/src/emotes/kitty.rs#L46 (which cause an infinite loop). Kitty icat feature seems to work throught tmux, but only on kitty, and seems broken on other terminals which support the graphics protocol.

I could maybe make it work but it would require an `TIOCGWINSZ` ioctl syscall to get the terminal size in pixel, which would introduce unsafe code. 

I have a last bug I need to fix. After a lot of emotes are loaded, if we resize the terminal it will take a long time to reload all emotes, and sometimes will crash. I'll try to find a way to either tell  ratatui to not clear the images, or maybe reload the images in another thread.

Nogesma commented:
Okay so I pushed an update which should fix the long freeze when resizing the window. I now only reload the emotes when needed.
I also switched to using the crossterm queues to send the graphics protocol commands, so it shouldn't flush the stdout buffer every command, it'll probably increase performance a bit.

Once this gets merge, I will probably make a PR to add this repo to the list of programs that support the graphics protocol on [kitty's website](https://sw.kovidgoyal.net/kitty/graphics-protocol/#terminal-graphics-protocol).

Nogesma commented:
Seems like I need to add some fallback function for versions of windows before 10. I don't think kitty is even supported on windows, and wezterm only supports windows 10 or later. I might just put a `panic!()` macro inside thoses functions if that's okay with you?

Xithrius commented:
Yep, alright with me.

Nogesma commented:
I didn't realize switching channels was a feature of twt. Switching channels should work with emotes now.

Nogesma commented:
Okay I think everything should be done now, let me know if I've missed anything.

balroggg commented:
Just checked emotes support from this branch, I have `foot` terminal emulator
```
cargo r --release
Running `target/release/twt`
The application panicked (crashed).
Message:  called `Result::unwrap()` on an `Err` value: 
   0: Configuration error.
   1: This terminal does not support the graphics protocol, please use a terminal such as Kitty or Wezterm, or disable emotes.

Location:
   src/handlers/config.rs:315
```
Maybe check terminal support by env vars and skip this feature for unsupported terminals?
For example:
```
// Check if the terminal is Kitty or not
fn check_kitty_support() -> bool {
    if let Ok(term) = std::env::var("TERM") {
        term.contains("kitty")
    } else {
        false
    }
}
```

Nogesma commented:
I'm doing exactly that [here](https://github.com/Nogesma/twitch-tui/blob/b7fe90cc55c8e3e990c5fa5424954d3c3bcb71da/src/emotes/graphics_protocol.rs#L342). Do you mean just disabling emotes silently if the terminal doesn't support it?

balroggg commented:
> I'm doing exactly that [here](https://github.com/Nogesma/twitch-tui/blob/b7fe90cc55c8e3e990c5fa5424954d3c3bcb71da/src/emotes/graphics_protocol.rs#L342). Do you mean just disabling emotes silently if the terminal doesn't support it?

Sorry I didn't notice this. Something like this, maybe return some kind of soft error(log the error and work or quit). Panic is the last thing users want to see) Just a suggestion, I know env vars maybe just been wrong.

kosayoda commented on src/emotes/graphics_protocol.rs:
```diff
@@ -0,0 +1,358 @@
+    stdout.write_all(command)?;
+    stdout.flush()?;
+
+    loop {
+        let c = stdout.read_key()?;
+        if let Key::UnknownEscSeq(_) = c {
+            break;
+        }
+    }
```
Can this function be documented/commented? For example, I'm not sure the reason for this loop, and whether or not there is a potential for an infinite loop. Would be nice to have an explanation for the response building code used below, or a link to the reference used.

Nogesma replied:
This can cause an infinite loop on terminals that do not respond to the <ESC>]14t/c queries, but it is only called if the terminal is kitty or wezterm, which do respond to theses queries. I read until I find an escape sequence because if the user start the program an inputs something, there might be some chars we need to read before we can read the réponse from the terminal. This is also why this function is called before we start the tui. 
If for some reason kitty/wezterm do not respond to the queries, it'll be an infinite loop. But I haven't found any other way. Maybe a readkey with timeout function exist?

kosayoda reviewed (changes requested):
Works well during manual testing, nice job.

Xithrius commented:
Sorry for noticing this only now, but I believe that more documentation describing how to set emotes up should be added to https://github.com/Xithrius/twitch-tui/blob/main/book/src/guide/configuration.md.

Nogesma commented on src/utils/pathing.rs:
```diff
@@ -18,6 +24,55 @@ pub fn config_path(file: &str) -> String {
+    }
+}
+
+pub fn create_temp_file(prefix: &str) -> Result<(File, PathBuf)> {
+    let (tempfile, pathbuf) = tempfile::Builder::new()
+        .prefix(prefix)
+        .rand_bytes(5)
+        .tempfile()?
+        .keep()?;
+
+    Ok((tempfile, pathbuf))
+}
```
Yes, kitty will actually delete the file once it has loaded the image.

Nogesma commented on default-config.toml:
```diff
@@ -7,6 +7,10 @@ channel = ""
 server = "irc.chat.twitch.tv"
 # OAuth token that authenticates you into Twitch. Acquired through https://twitchapps.com/tmi/.
 token = ""
+# API Token used to download emotes. Acquired throught https://twitchtokengenerator.com.
+api = ""
+# Client-Id used to authenticate with the api. Acquired throught https://twitchtokengenerator.com.
+id = ""
```
I've tested it and I don't even think we need any scopes for the token. I'm only using it to get a channel id from the channel name, and then to download the channel twitch emotes.

Nogesma replied:
Actually I just realized I can just reuse the irc token.

Xithrius reviewed (approved):
Looks good, works perfectly with just the IRC token.

kosayoda reviewed (approved):
Should be good to merge, any issues can be fixed in a separate PR. Thanks!

Xithrius commented:
Thank you so much for the PR! I'll get a release going soon.
